apps/showcase/src/services/webcontainer/webcontainer.service.ts

```diff
@@ -0,0 +1,133 @@
+import {Injectable} from '@angular/core';
+import {FileSystem, getFilesTree} from '@o3r-training/tools';
+import {DirectoryNode, FileNode, FileSystemTree, WebContainer} from '@webcontainer/api';
+import {BehaviorSubject, distinctUntilChanged, map} from 'rxjs';
+import {MonacoTreeElement} from 'ngx-monaco-tree';
+import {WebContainerRunner} from './webcontainer-runner';
+
+/** List of files or directories to exclude from the file tree */
+const EXCLUDED_FILES_OR_DIRECTORY = ['node_modules', '.angular', '.vscode'];
+
+/**
+ * Convert the given path and node to a MonacoTreeElement
+ * @param path
+ * @param node
+ */
+function convertTreeRec(path: string, node: DirectoryNode | FileNode): MonacoTreeElement {
+  return {
+    name: path,
+    content: (node as DirectoryNode).directory
+      ? Object.entries((node as DirectoryNode).directory)
+        .map(([p, n]) => convertTreeRec(p, n))
+      : undefined
+  };
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class WebcontainerService {
+  public readonly runner = new WebContainerRunner();
+  private readonly monacoTree = new BehaviorSubject<MonacoTreeElement[]>([]);
+
+  public monacoTree$ = this.monacoTree.pipe(
+    distinctUntilChanged((a, b) => JSON.stringify(a) === JSON.stringify(b))
+  );
+  public isReady$ = this.monacoTree$.pipe(
+    map((tree) => tree.length > 0)
+  );
+
+  /**
+   * Get the Monaco file tree from the given root path
+   * @param rootPath
+   * @private
+   */
+  private async getMonacoTree(rootPath: string): Promise<MonacoTreeElement[]> {
+    const instance = await this.runner.instance;
+    // eslint-disable-next-line no-console
+    console.log('Monaco - start', rootPath);
+    const tree = await this.getFilesTreeFromContainer(instance, rootPath);
+    // eslint-disable-next-line no-console
+    console.log('Monaco - ', rootPath, tree);
+    return Object.entries(tree)
+      .reduce(
+        (acc: MonacoTreeElement[], [path, node]) =>
+          acc.concat(convertTreeRec(path, node)),
+        []
+      );
```

Comment:
any reason to use a reduce instead of just a map ?
```suggestion
      .map(([path, node]) => convertTreeRec(path, node));
```